src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/DiversityBuilder.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.sampler;
+
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.aggregations.ValuesSourceAggregationBuilder;
+
+import java.io.IOException;
+
+public class DiversityBuilder extends ValuesSourceAggregationBuilder<DiversityBuilder> {
+
+    int maxDocsPerKey = SamplerParser.MAX_DOCS_PER_KEY_DEFAULT;
+    String executionHint = null;
+
+    protected DiversityBuilder() {
+        super("diversify", "diversify");
+    }
+
+    public DiversityBuilder maxDocsPerKey(int maxDocsPerKey) {
+        this.maxDocsPerKey = maxDocsPerKey;
+        return this;
+    }
+
+    public DiversityBuilder executionHint(String executionHint) {
+        this.executionHint = executionHint;
+        return this;
+    }
+
+    @Override
+    protected XContentBuilder doInternalXContent(XContentBuilder builder, Params params) throws IOException {
+        if (maxDocsPerKey != SamplerParser.MAX_DOCS_PER_KEY_DEFAULT) {
+            builder.field(SamplerParser.MAX_DOCS_PER_KEY_FIELD.getPreferredName(), maxDocsPerKey);
+        }
+        if (executionHint != null) {
+            builder.field(SamplerParser.EXECUTION_HINT_FIELD.getPreferredName(), executionHint);
+        }
+        return builder;
+    }
+
+    public void exposeInternalXContent(XContentBuilder builder, Params params) throws IOException {
```

Comment:
Actually I kind of like option 3 as then we would have a ValuesSourceBuilder to go with the ValuesSourceParser. I don't think the parser side relies on being nested at the same level as the agg so the builder side shouldn't either. I also think that this way we wouldn't need to have the DiversityBuilder which would be nice since there is no DiversityParser on the other side (all the parsing is done by the SamplerParser) The ValuesSourceAggregationBuilder and SamplerBuilder can just use rather than extend the ValuesSourceBuilder. WDYT?